elastic/kibana
# [SecuritySolution][EntityAnalytics] Account for Asset Criticality in Risk Scoring

## Summary

This PR adds asset criticality to the risk scoring calculation. While previous work was done to add the resources and client to create/read asset criticality documents, this PR builds on those to retrieve relevant criticality records and apply them to the intermediate scores calculated via elasticsearch.

__Note that the code in this PR that performs the actual calculation/writing of new risk fields is behind a feature flag.__

### Performance
Since this PR adds new logic to an already tight/important code path, we need to ensure that we haven't changed behavior nor performance. I've captured that as a separate task to be done next: https://github.com/elastic/security-team/issues/8223. 

### Compatibility
 Behaviorally, without the feature flag enabled, scoring will skip the criticality workflow and not write new fields (`criticality_level`, `criticality_modifier`). ~~I still have an outstanding task to validate whether this will be an issue until criticality is enabled, and do some smarter short-circuiting.~~ This task uncovered our need for the above feature flag.

The one behavioral change introduced in this PR _not_ behind a feature flag is the normalization of our risk category scores.

### Summary of Changes
- Adds an `AssetCriticalityService` which provides the API used by risk scoring to retrieve criticality records
-  Adds new `search` method to the `AssetCriticalityDataClient`: used to generally retrieve multiple criticality records at once
-  Adds functions to calculate a (currently hard-coded) modifier from a criticality level, and apply it to the risk score via bayesian update
-  Moves risk score level calculation into javascript (necessary because we need the level to account for criticality)
-  Moves some risk level code out of threat hunting code and into the `common/entity_analytics` folder
-  Tests and comments throughout


### TODO
- [x] Add new criticality fields to risk score, address upgrade workflow
- [x] Validate that code works without criticality being enabled.
- [x] Bump task version to invalidate old tasks from being picked up in customer env

Outcome: All three of the above are addressed with:

1. Moving the code responsible for adding new fields behind a feature flag ([71f1158](https://github.com/elastic/kibana/pull/172417/commits/71f115800b9ecca21d1c2e0e9014163f2858ece3))
2. Addressing the upgrade path in a subsequent issue (https://github.com/elastic/security-team/issues/8012)

## How to Review
1. Enable our asset criticality feature flag:
> ```
> xpack.securitySolution.enableExperimental: 
>  - entityAnalyticsAssetCriticalityEnabled
> ```
2. Create asset criticality records (see API documentation, or follow [this test](https://github.com/elastic/kibana/pull/172417/files#diff-43f9f394fb7c8eb0f0ace3f5e75482c56a7233ae7d11d5fdb98a89e6404412c3R276) as a setup guide
3. Enable risk engine
4. Observe that new fields are written to risk scores' `_source`, but not mapped/searchable
5. (optional) Observe that the transform subsequently fails 😢 


### Checklist
